_build/lexicon/checklexicon.php

```diff
@@ -0,0 +1,565 @@
+                        $pathinfo['extension'] == 'tpl' ||
+                        $pathinfo['basename'] == 'config.json'
+                    ) &&
+                    strpos($pathinfo['basename'], 'min.js') === false &&
+                    strpos($pathinfo['basename'], 'ext-') !== 0
+                ) ? true : false;
+            }
+        });
+        $iterator = new \RecursiveIteratorIterator($filter);
+
+        foreach ($iterator as $path => $current) {
+            $this->addPhpKeys($path);
+            $this->addJsKeys($path);
+            $this->addChunkKeys($path);
+            $this->addSmartyKeys($path);
+        }
+        $this->addSettingKeys();
+        $this->addMenuKeys();
+        $this->addWidgetKeys();
+        $this->addPermissionKeys();
+
+        $this->languageKeys = array_unique($this->languageKeys);
+        sort($this->languageKeys);
+    }
+
+    /**
+     * Add lexicon calls in php files:
+     * modx->lexicon('whatever'
+     *
+     * @param string $filename
+     */
+    private function addPhpKeys($filename)
+    {
+        $fileContent = file_get_contents($filename);
+        $results = array();
+        preg_match_all('/(modx|xpdo)->lexicon\((["\'])(.*?)\2\s*[,\)]/m', $fileContent, $results);
+        if (is_array($results[3])) {
+            foreach ($results[3] as $result) {
+                // Don't add lexicon keys that ends with a dot or an underscore or that contain a variable
+                if (substr($result, -1) !== '.' &&
+                    substr($result, -1) !== '_'
+                ) {
+                    if (strpos($result, '$') === false
+                    ) {
+                        $this->languageKeys[] = $result;
+                    } else {
+                        $this->variableKeys[] = $result;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Add lexicon calls in javascript files:
+     * _('whatever'
+     * createDelegate(this, ['whatever'
+     *
+     * @param string $filename
+     */
+    private function addJsKeys($filename)
+    {
+        $fileContent = file_get_contents($filename);
+        $results = array();
+        preg_match_all('/_\(([\'"])(.*?)\1\s*[,\)]/m', $fileContent, $results);
+        if (is_array($results[2])) {
+            foreach ($results[2] as $result) {
+                // Don't add lexicon keys that ends with a dot or an underscore or that key is concatenated
+                if (substr($result, -1) !== '.' &&
+                    substr($result, -1) !== '_'
+                ) {
+                    if (strpos($result, '+') === false
+                    ) {
+                        $this->languageKeys[] = $result;
+                    } else {
+                        $this->variableKeys[] = $result;
+                    }
+                }
+            }
+        }
+        preg_match_all('/(createDelegate)\(.*?, \[([\'"])(.*?)\2/m', $fileContent, $results);
```

Comment:
We should consider using [Named Capturing Groups](https://www.regular-expressions.info/named.html) here.